### ISSUE TITLE
Issues with individual nodes services startup order (e.g. Wireguard internal IPs ready after the services that depend on it)

### ISSUE THREAD

**Issue by fititnt** [fititnt/ansible-linux-ha-cluster#4]
_**TL;DR: when using Wireguard in production any software started on boot that may fail to bind a private IP, bug when you inspect is everything working, may be a symptom of your software starting BEFORE Wireguard is ready; this may be as low as 1~2s or as long as 2min timeout of misconfigured systemd-networkd-wait-online.service (used by Wireguard to detect if it can start)**_

----

Maybe related:
- <https://wiki.archlinux.org/index.php/Talk:WireGuard#systemd-networkd-wait-online.service_needed?>
- <https://lists.zx2c4.com/pipermail/wireguard/2018-February/002360.html>
- <https://serverfault.com/questions/972634/debian-slow-boot-waiting-for-network>
- https://github.com/systemd/systemd/issues/6441
- https://github.com/systemd/systemd/issues/5154

Other references
- http://man7.org/linux/man-pages/man8/systemd-networkd-wait-online.service.8.html

---

### Pratical example: 

While testing an stack similar to lastest one from this repository on a private project in a cluster on Contabo (but the way IPv6 is configured may also affect any other service), more specifically:
- 18.04.3 LTS (Bionic Beaver)
  - `systemd --version: systemd 237`
- wireguard
- openresty (bind do wireguard internal IP) 
- haproxy
- Consul (in cluster)
- MariaDB (galera cluster)

The services that depends on internal IPs and did not automatically retry (like the Consul) may fail at reboot and need human intervention (on this setup rebooted nodes of Galera Cluster and openresty) to try start again the services.

Since I'm doing this ansible-linux-ha-cluster repo, I will take some extra time do eventually document workarounds for this. And go this deep actually may take more time to debug than I supposed at first. No pain, no gain.

---
Notes:
- This issue is unlikely to affect people who does not depend on clustered software or software that does not bind at one Wireguard port and doesn't retry. Also is not granteed that other OSs and other VPS providers may have the same issue.
- This issue may, actually, not be blammed on a single software, but a combo of conditions. Wireguard is very, very fast, and much more easy to configure than other alternatives

**Comment by fititnt** [comment]
Here the original service files for the relevant services.

At least on Ubuntu, the ppa from `githubixx.ansible_role_wireguard` by default wnat to be loaded `After=network-online.target nss-lookup.target`, while most of other packages will simply ask `After=network.target`.

Systemd try it's best to decide, based on dependencies and etc, what to load first and to it asynchronously (that why very fast even with lots of services).

On the private project I tried to just customize the wg-quick@.service to add itself as dependency for the other services (using variables like `wireguard_systemd_before`, `wireguard_systemd_wanted_by` and `wireguard_systemd_required_by`, but maybe it didin't work because `network-online.target` depends on `network.target`.

Humm...

## Files content


### /lib/systemd/system/wg-quick@.service

```bash
[Unit]
Description=WireGuard via wg-quick(8) for %I
After=network-online.target nss-lookup.target
Wants=network-online.target nss-lookup.target
Documentation=man:wg-quick(8)
Documentation=man:wg(8)
Documentation=https://www.wireguard.com/
Documentation=https://www.wireguard.com/quickstart/
Documentation=https://git.zx2c4.com/wireguard-tools/about/src/man/wg-quick.8
Documentation=https://git.zx2c4.com/wireguard-tools/about/src/man/wg.8

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/usr/bin/wg-quick up %i
ExecStop=/usr/bin/wg-quick down %i
Environment=WG_ENDPOINT_RESOLUTION_RETRIES=infinity

[Install]
WantedBy=multi-user.target

```

### /lib/systemd/system/haproxy.service

```bash
[Unit]
Description=HAProxy Load Balancer
Documentation=man:haproxy(1)
Documentation=file:/usr/share/doc/haproxy/configuration.txt.gz
After=network.target rsyslog.service

[Service]
EnvironmentFile=-/etc/default/haproxy
EnvironmentFile=-/etc/sysconfig/haproxy
Environment="CONFIG=/etc/haproxy/haproxy.cfg" "PIDFILE=/run/haproxy.pid" "EXTRAOPTS=-S /run/haproxy-master.sock"
ExecStartPre=/usr/sbin/haproxy -f $CONFIG -c -q $EXTRAOPTS
ExecStart=/usr/sbin/haproxy -Ws -f $CONFIG -p $PIDFILE $EXTRAOPTS
ExecReload=/usr/sbin/haproxy -f $CONFIG -c -q $EXTRAOPTS
ExecReload=/bin/kill -USR2 $MAINPID
KillMode=mixed
Restart=always
SuccessExitStatus=143
Type=notify

# The following lines leverage SystemD's sandboxing options to provide
# defense in depth protection at the expense of restricting some flexibility
# in your setup (e.g. placement of your configuration files) or possibly
# reduced performance. See systemd.service(5) and systemd.exec(5) for further
# information.

# NoNewPrivileges=true
# ProtectHome=true
# If you want to use 'ProtectSystem=strict' you should whitelist the PIDFILE,
# any state files and any other files written using 'ReadWritePaths' or
# 'RuntimeDirectory'.
# ProtectSystem=true
# ProtectKernelTunables=true
# ProtectKernelModules=true
# ProtectControlGroups=true
# If your SystemD version supports them, you can add: @reboot, @swap, @sync
# SystemCallFilter=~@cpu-emulation @keyring @module @obsolete @raw-io

[Install]
WantedBy=multi-user.target

```

### /lib/systemd/system/openresty.service

```bash
# Stop dance for OpenResty
# =========================
#
# ExecStop sends SIGSTOP (graceful stop) to OpenResty's nginx process.
# If, after 5s (--retry QUIT/5) nginx is still running, systemd takes control
# and sends SIGTERM (fast shutdown) to the main process.
# After another 5s (TimeoutStopSec=5), and if nginx is alive, systemd sends
# SIGKILL to all the remaining processes in the process group (KillMode=mixed).
#
# nginx signals reference doc:
# http://nginx.org/en/docs/control.html
#
[Unit]
Description=full-fledged web platform
After=network.target

[Service]
Type=forking
PIDFile=/usr/local/openresty/nginx/logs/nginx.pid
ExecStartPre=/usr/local/openresty/nginx/sbin/nginx -t -q -g 'daemon on; master_process on;'
ExecStart=/usr/local/openresty/nginx/sbin/nginx -g 'daemon on; master_process on;'
ExecReload=/usr/local/openresty/nginx/sbin/nginx -g 'daemon on; master_process on;' -s reload
ExecStop=-/sbin/start-stop-daemon --quiet --stop --retry QUIT/5 --pidfile /usr/local/openresty/nginx/logs/nginx.pid
TimeoutStopSec=5
KillMode=mixed

[Install]
WantedBy=multi-user.target

```

### /lib/systemd/system/mariadb.service

```bash
# It's not recommended to modify this file in-place, because it will be
# overwritten during package upgrades.  If you want to customize, the
# best way is to create a file "/etc/systemd/system/mariadb.service",
# containing
#	.include /usr/lib/systemd/system/mariadb.service
#	...make your changes here...
# or create a file "/etc/systemd/system/mariadb.service.d/foo.conf",
# which doesn't need to include ".include" call and which will be parsed
# after the file mariadb.service itself is parsed.
#
# For more info about custom unit files, see systemd.unit(5) or
# https://mariadb.com/kb/en/mariadb/systemd/
#
# Copyright notice:
#
# This file is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or
# (at your option) any later version.

[Unit]
Description=MariaDB 10.4.10 database server
Documentation=man:mysqld(8)
Documentation=https://mariadb.com/kb/en/library/systemd/
After=network.target

[Install]
WantedBy=multi-user.target
Alias=mysql.service
Alias=mysqld.service


[Service]

##############################################################################
## Core requirements
##

Type=notify

# Setting this to true can break replication and the Type=notify settings
# See also bind-address mysqld option.
PrivateNetwork=false

##############################################################################
## Package maintainers
##

User=mysql
Group=mysql

# CAP_IPC_LOCK To allow memlock to be used as non-root user
# CAP_DAC_OVERRIDE To allow auth_pam_tool (which is SUID root) to read /etc/shadow when it's chmod 0
#   does nothing for non-root, not needed if /etc/shadow is u+r
# CAP_AUDIT_WRITE auth_pam_tool needs it on Debian for whatever reason
CapabilityBoundingSet=CAP_IPC_LOCK CAP_DAC_OVERRIDE CAP_AUDIT_WRITE

# PrivateDevices=true implies NoNewPrivileges=true and
# SUID auth_pam_tool suddenly doesn't do setuid anymore
PrivateDevices=false

# Prevent writes to /usr, /boot, and /etc
ProtectSystem=full

# Doesn't yet work properly with SELinux enabled
# NoNewPrivileges=true

# Prevent accessing /home, /root and /run/user
ProtectHome=true

# Execute pre and post scripts as root, otherwise it does it as User=
PermissionsStartOnly=true

ExecStartPre=/usr/bin/install -m 755 -o mysql -g root -d /var/run/mysqld

# Perform automatic wsrep recovery. When server is started without wsrep,
# galera_recovery simply returns an empty string. In any case, however,
# the script is not expected to return with a non-zero status.
# It is always safe to unset _WSREP_START_POSITION environment variable.
# Do not panic if galera_recovery script is not available. (MDEV-10538)
ExecStartPre=/bin/sh -c "systemctl unset-environment _WSREP_START_POSITION"
ExecStartPre=/bin/sh -c "[ ! -e /usr/bin/galera_recovery ] && VAR= || \
 VAR=`/usr/bin/galera_recovery`; [ $? -eq 0 ] \
 && systemctl set-environment _WSREP_START_POSITION=$VAR || exit 1"

# Needed to create system tables etc.
# ExecStartPre=/usr/bin/mysql_install_db -u mysql

# Start main service
# MYSQLD_OPTS here is for users to set in /etc/systemd/system/mariadb.service.d/MY_SPECIAL.conf
# Use the [Service] section and Environment="MYSQLD_OPTS=...".
# This isn't a replacement for my.cnf.
# _WSREP_NEW_CLUSTER is for the exclusive use of the script galera_new_cluster
ExecStart=/usr/sbin/mysqld $MYSQLD_OPTS $_WSREP_NEW_CLUSTER $_WSREP_START_POSITION

# Unset _WSREP_START_POSITION environment variable.
ExecStartPost=/bin/sh -c "systemctl unset-environment _WSREP_START_POSITION"

ExecStartPost=/etc/mysql/debian-start

KillSignal=SIGTERM

# Don't want to see an automated SIGKILL ever
SendSIGKILL=no

# Restart crashed server only, on-failure would also restart, for example, when
# my.cnf contains unknown option
Restart=on-abort
RestartSec=5s

UMask=007

##############################################################################
## USERs can override
##
##
## by creating a file in /etc/systemd/system/mariadb.service.d/MY_SPECIAL.conf
## and adding/setting the following under [Service] will override this file's
## settings.

# Useful options not previously available in [mysqld_safe]

# Kernels like killing mysqld when out of memory because its big.
# Lets temper that preference a little.
# OOMScoreAdjust=-600

# Explicitly start with high IO priority
# BlockIOWeight=1000

# If you don't use the /tmp directory for SELECT ... OUTFILE and
# LOAD DATA INFILE you can enable PrivateTmp=true for a little more security.
PrivateTmp=false

##
## Options previously available to be set via [mysqld_safe]
## that now needs to be set by systemd config files as mysqld_safe
## isn't executed.
##

# Number of files limit. previously [mysqld_safe] open-file-limit
LimitNOFILE=16364

# Maximium core size. previously [mysqld_safe] core-file-size
# LimitCore=

# Nice priority. previously [mysqld_safe] nice
# Nice=-5

# Timezone. previously [mysqld_safe] timezone
# Environment="TZ=UTC"

# Library substitutions. previously [mysqld_safe] malloc-lib with explicit paths
# (in LD_LIBRARY_PATH) and library name (in LD_PRELOAD).
# Environment="LD_LIBRARY_PATH=/path1 /path2" "LD_PRELOAD=

# Flush caches. previously [mysqld_safe] flush-caches=1
# ExecStartPre=sync
# ExecStartPre=sysctl -q -w vm.drop_caches=3

# numa-interleave=1 equalivant
# Change ExecStart=numactl --interleave=all /usr/sbin/mysqld......

# crash-script equalivent
# FailureAction=

```

----

Edit: added example of how is loaded without changes

```txt
root@servername:~# journalctl -b -u wg-quick@wg0.service -u haproxy.service -u openresty.service -u mariadb.service -u systemd-networkd-wait-online.service -u mariadb.service
-- Logs begin at Tue 2020-01-28 17:04:38 UTC, end at Sat 2020-02-29 02:50:09 UTC. --
Feb 29 02:37:56 servername systemd[1]: Starting Wait for Network to be Configured...
Feb 29 02:37:56 servername systemd-networkd-wait-online[339]: ignoring: lo
Feb 29 02:37:56 servername systemd-networkd-wait-online[339]: ignoring: lo
Feb 29 02:37:57 servername systemd[1]: Starting full-fledged web platform...
Feb 29 02:37:57 servername systemd[1]: Starting HAProxy Load Balancer...
Feb 29 02:37:57 servername systemd[1]: Starting MariaDB 10.4.10 database server...
Feb 29 02:37:57 servername haproxy[595]: [WARNING] 059/023757 (595) : parsing [/etc/haproxy/haproxy.cfg:77] : backend 'backend_http' : 'option tcplog' directive is ignored in backends.
Feb 29 02:37:57 servername haproxy[595]: [WARNING] 059/023757 (595) : parsing [/etc/haproxy/haproxy.cfg:53] : 'option httplog' not usable with proxy 'galera' (needs 'mode http'). Falling back to 'option tcplog'.
Feb 29 02:37:57 servername haproxy[595]: Proxy frontend_http started.
Feb 29 02:37:57 servername haproxy[595]: Proxy frontend_http started.
Feb 29 02:37:57 servername haproxy[595]: Proxy backend_http started.
Feb 29 02:37:57 servername haproxy[595]: Proxy backend_http started.
Feb 29 02:37:57 servername haproxy[595]: Proxy https-in started.
Feb 29 02:37:57 servername haproxy[595]: Proxy https-in started.
Feb 29 02:37:57 servername haproxy[595]: Proxy stats started.
Feb 29 02:37:57 servername haproxy[595]: Proxy stats started.
Feb 29 02:37:57 servername haproxy[595]: Proxy galera started.
Feb 29 02:37:57 servername haproxy[595]: Proxy galera started.
Feb 29 02:37:57 servername haproxy[595]: [NOTICE] 059/023757 (595) : New worker #1 (597) forked
Feb 29 02:37:57 servername systemd[1]: Started HAProxy Load Balancer.
Feb 29 02:37:57 servername haproxy[595]: [WARNING] 059/023757 (597) : Server backend_http/saturno_bravo is DOWN, reason: Layer4 connection problem, info: "General socket error (Network is unreachable)", check duration: 0ms. 0 act
Feb 29 02:37:57 servername haproxy[595]: [ALERT] 059/023757 (597) : backend 'backend_http' has no server available!
Feb 29 02:37:57 servername nginx[580]: nginx: [emerg] bind() to 10.8.0.103:8080 failed (99: Cannot assign requested address)
Feb 29 02:37:57 servername nginx[580]: nginx: configuration file /usr/local/openresty/nginx/conf/nginx.conf test failed
Feb 29 02:37:57 servername systemd[1]: openresty.service: Control process exited, code=exited status=1
Feb 29 02:37:57 servername systemd[1]: openresty.service: Failed with result 'exit-code'.
Feb 29 02:37:57 servername systemd[1]: Failed to start full-fledged web platform.
Feb 29 02:37:57 servername haproxy[595]: [WARNING] 059/023757 (597) : Server galera/odin_bravo is DOWN, reason: Layer4 connection problem, info: "General socket error (Network is unreachable)", check duration: 0ms. 2 active and 0
Feb 29 02:37:58 servername systemd-networkd-wait-online[339]: ignoring: lo
Feb 29 02:37:58 servername systemd-networkd-wait-online[339]: managing: eth0
Feb 29 02:37:58 servername systemd[1]: Started Wait for Network to be Configured.
Feb 29 02:37:58 servername systemd[1]: Starting WireGuard via wg-quick(8) for wg0...
Feb 29 02:37:58 servername wg-quick[800]: [#] ip link add wg0 type wireguard
Feb 29 02:37:58 servername wg-quick[800]: [#] wg setconf wg0 /dev/fd/63
Feb 29 02:37:59 servername wg-quick[800]: [#] ip -4 address add 10.8.0.103/24 dev wg0
Feb 29 02:37:59 servername wg-quick[800]: [#] ip link set mtu 1420 up dev wg0
Feb 29 02:37:59 servername systemd[1]: Started WireGuard via wg-quick(8) for wg0.
Feb 29 02:38:00 servername haproxy[595]: [WARNING] 059/023800 (597) : Server galera/thor_bravo is DOWN, reason: Layer4 timeout, check duration: 2000ms. 1 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remainin
Feb 29 02:38:00 servername haproxy[595]: [WARNING] 059/023800 (597) : Server galera/saturno_bravo is DOWN, reason: Socket error, info: "Connection reset by peer", check duration: 1642ms. 0 active and 0 backup servers left. 0 sess
Feb 29 02:38:00 servername haproxy[595]: [ALERT] 059/023800 (597) : proxy 'galera' has no server available!
Feb 29 02:38:00 servername mysqld[605]: WSREP: Recovered position e0a37db6-41fa-11ea-9f74-4e8ff6068297:79348
Feb 29 02:38:00 servername mysqld[1348]: 2020-02-29  2:38:00 0 [Note] /usr/sbin/mysqld (mysqld 10.4.10-MariaDB-1:10.4.10+maria~bionic-log) starting as process 1348 ...
(...)

```
